src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/Native/Interop.crypt32.cs

```diff
@@ -89,8 +89,33 @@ IntPtr ppvContext
         [DllImport(Libraries.Crypt32, CharSet = CharSet.Unicode, SetLastError = true)]
         public static extern unsafe bool CertSetCertificateContextProperty(SafeCertContextHandle pCertContext, CertContextPropId dwPropId, CertSetPropertyFlags dwFlags, [In] SafeNCryptKeyHandle keyHandle);
 
+        public static unsafe string CertGetNameString(
+            SafeCertContextHandle certContext,
+            CertNameType certNameType,
+            CertNameFlags certNameFlags,
+            int strType) // CertNameStringType or CertNameStrTypeAndFlags
+        {
+            int cchCount = CertGetNameString(certContext, certNameType, certNameFlags, strType, null, 0);
+            if (cchCount == 0)
+            {
+                throw Marshal.GetLastWin32Error().ToCryptographicException();
+            }
+
+            Span<char> buffer = cchCount <= 128 ? stackalloc char[cchCount] : new char[cchCount];
+            fixed (char* ptr = &MemoryMarshal.GetReference(buffer))
+            {
+                if (CertGetNameString(certContext, certNameType, certNameFlags, strType, ptr, cchCount) == 0)
+                {
+                    throw Marshal.GetLastWin32Error().ToCryptographicException();
+                }
+
+                Debug.Assert(buffer[cchCount - 1] == '\0');
```

Comment:
@bartonjs, is null termination actually guaranteed here or could the cert decide to skip it?

Reply:
It's guaranteed (see https://msdn.microsoft.com/en-us/library/windows/desktop/aa376086(v=vs.85).aspx), and if it weren't StringBuilder marshaling would be falling off a cliff.